cmd/thanos/sidecar.go

```diff
@@ -209,14 +209,41 @@ func runSidecar(
 			cancel()
 		})
 	}
+	{
+		ctx, cancel := context.WithCancel(context.Background())
+		g.Add(func() error {
+			// We retry infinitely until we reach and fetch BuildVersion from our Prometheus.
+			err := runutil.Retry(2*time.Second, ctx.Done(), func() error {
+				if err := m.BuildVersion(ctx); err != nil {
+					level.Warn(logger).Log(
+						"msg", "failed to fetch prometheur version. Is Prometheus running? Retrying",
+						"err", err,
+					)
+					return err
+				}
```

Comment:
I am still thinking about this logic.
This doesn't work well when Prometheus version is < 2.14.0, where `/api/v1/status/buildinfo` doesn't exist. In this case, we just get 404s but we are still retrying indefinitely. This seems wrong to me.

We can only retry for a short time or we check the return error to stop retrying.

Reply:
Lets retry for lets say `N` times (maybe `N=5`) and exit if retry fails?

Reply:
Also for Prometheus `<2.14.0` this function would fail, so we also need to decide how to fetch build version for `<2.14.0`

Reply:
As discussed in contributor hours:
Let’s check for 404, if it’s present, set 0 value for promVersion